src/coreComponents/constitutive/fluid/PVTFunctions/SpanWagnerCO2DensityFunction.cpp

```diff
@@ -120,217 +133,97 @@ real64 f( const real64 & T, const real64 & P, const real64 & rho )
-
-  dT = -1.0;
-  dP = -1.0;
-  TStart = -1.0;
-  TEnd = -1.0;
-  PStart = -1.0;
-  PEnd = -1.0;
-
-  GEOSX_ERROR_IF( inputPara.size() < 8, "Invalid SpanWagnerCO2Density input!" );
-
-  try
-  {
-
-    PStart = stod( inputPara[2] );
-    PEnd = stod( inputPara[3] );
-    dP = stod( inputPara[4] );
-
-    TStart = stod( inputPara[5] );
-    TEnd = stod( inputPara[6] );
-    dT = stod( inputPara[7] );
-
-  }
-  catch( const std::invalid_argument & e )
-  {
-
-    GEOSX_ERROR( "Invalid SpanWagnerCO2Density argument:" + string( e.what()));
-
-  }
-
-  P = PStart;
-
-  while( P <= PEnd )
-  {
-
-    pressures.emplace_back( P );
-    P += dP;
-
-  }
-
-  T = TStart;
-
-  while( T <= TEnd )
-  {
-
-    temperatures.emplace_back( T );
-    T += dT;
-
-  }
-
-  localIndex nP = pressures.size();
-  localIndex nT = temperatures.size();
-
-  real64_array2d densities( nP, nT );
-
-  calculateCO2Density( pressures, temperatures, densities );
-
-  m_CO2DensityTable = std::make_shared< XYTable >( "SpanWagnerCO2DensityTable", pressures, temperatures, densities );
-
-
+  PTTableCoordinates tableCoords;
+  PVTFunctionHelpers::initializePropertyTable( inputPara, tableCoords );
+
+  array1d< real64 > densities( tableCoords.nPressures() * tableCoords.nTemperatures() );
+  calculateCO2Density( tableCoords, densities );
+
+  FunctionManager & functionManager = getGlobalState().getFunctionManager();
+  m_CO2DensityTable = dynamicCast< TableFunction * >( functionManager.createChild( "TableFunction", "CO2DensityTable" ) );
+  m_CO2DensityTable->setTableCoordinates( tableCoords.getCoords() );
+  m_CO2DensityTable->setTableValues( densities );
+  m_CO2DensityTable->reInitializeFunction();
+  m_CO2DensityTable->setInterpolationMethod( TableFunction::InterpolationType::Linear );
 }
 
 
-void SpanWagnerCO2DensityFunction::evaluation( EvalVarArgs const & pressure, EvalVarArgs const & temperature, arraySlice1d< EvalVarArgs const > const & GEOSX_UNUSED_PARAM(
-                                                 phaseComposition ), EvalVarArgs & value, bool useMass ) const
-{
-
-  EvalArgs2D P, T, density;
-  P.m_var = pressure.m_var;
-  P.m_der[0] = 1.0;
-
-  T.m_var = temperature.m_var;
-  T.m_der[1] = 1.0;
-
-  density = m_CO2DensityTable->value( P, T );
-
-  real64 CO2MW = m_componentMolarWeight[m_CO2Index];
-
-  if( !useMass )
-  {
-
-    density /= CO2MW;
-
-  }
-
-  value.m_var = density.m_var;
-  value.m_der[0] = density.m_der[0];
-
-}
-
-void SpanWagnerCO2DensityFunction::calculateCO2Density( real64_array const & pressure, real64_array const & temperature, real64_array2d const & density )
+void SpanWagnerCO2Density::calculateCO2Density( PTTableCoordinates const & tableCoords,
+                                                array1d< real64 > const & densities )
 {
 
-  constexpr real64 T_K_f = 273.15;
-
-  real64 PPa, TK;
+  constexpr real64 TK_f = 273.15;
 
-  for( localIndex i = 0; i < pressure.size(); ++i )
+  for( localIndex i = 0; i < tableCoords.nPressures(); ++i )
```

Comment:
Same here, please take the `nTemperatures` and `nPressures` out of the loops.

Reply:
Done